File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/widgets/WidgetChooseImage.java
package com.sup.dev.android.views.widgets;

import android.database.Cursor;
import android.graphics.Bitmap;
import android.provider.MediaStore;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.widget.GridLayoutManager;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.sup.dev.android.R;
import com.sup.dev.android.app.SupAndroid;
import com.sup.dev.android.libs.image_loader.ImageLoader;
import com.sup.dev.android.libs.image_loader.ImageLoaderFile;
import com.sup.dev.android.tools.ToolsAndroid;
import com.sup.dev.android.tools.ToolsBitmap;
import com.sup.dev.android.tools.ToolsFiles;
import com.sup.dev.android.tools.ToolsPermission;
import com.sup.dev.android.tools.ToolsResources;
import com.sup.dev.android.tools.ToolsToast;
import com.sup.dev.android.tools.ToolsView;
import com.sup.dev.android.views.adapters.recycler_view.RecyclerCardAdapter;
import com.sup.dev.android.views.cards.Card;
import com.sup.dev.android.views.dialogs.DialogSheetWidget;
import com.sup.dev.android.views.dialogs.DialogWidget;
import com.sup.dev.android.views.views.ViewIcon;
import com.sup.dev.java.classes.callbacks.simple.Callback2;
import com.sup.dev.java.libs.debug.Debug;
import com.sup.dev.java.tools.ToolsBytes;
import com.sup.dev.java.tools.ToolsNetwork;

import java.io.File;
import java.io.IOException;

public class WidgetChooseImage extends WidgetRecycler {

    private final RecyclerCardAdapter adapter;

    private Callback2<WidgetChooseImage, byte[]> onSelected;
    private boolean imagesLoaded;

    public WidgetChooseImage() {
        adapter = new RecyclerCardAdapter();
        View vFabGalleryContainer = ToolsView.inflate(R.layout.view_fab);
        View vFabLinkContainer = ToolsView.inflate(R.layout.view_fab);
        FloatingActionButton vFabGallery = vFabGalleryContainer.findViewById(R.id.fab);
        FloatingActionButton vFabLink = vFabLinkContainer.findViewById(R.id.fab);
        vContainer.addView(vFabGalleryContainer);
        vContainer.addView(vFabLinkContainer);

        ((ViewGroup.MarginLayoutParams) vFabLinkContainer.getLayoutParams()).rightMargin = ToolsView.dpToPx(72);

        vRecycler.setLayoutManager(new GridLayoutManager(view.getContext(), ToolsAndroid.isScreenPortrait() ? 3 : 6));

        vFabGallery.setImageResource(R.drawable.ic_landscape_white_24dp);
        vFabLink.setImageResource(R.drawable.ic_insert_link_white_24dp);
        vFabGallery.setOnClickListener(v -> openGallery());
        vFabLink.setOnClickListener(v -> showLink());

        setAdapter(adapter);
    }

    @Override
    public void onShow() {
        super.onShow();
        loadImages();

        ((ViewGroup.MarginLayoutParams) vRecycler.getLayoutParams()).setMargins(0, ToolsView.dpToPx(2), 0, 0);
        vRecycler.getLayoutParams().height = ViewGroup.LayoutParams.MATCH_PARENT;

        if (viewWrapper instanceof DialogWidget)
            ((ViewGroup.MarginLayoutParams) vRecycler.getLayoutParams()).setMargins(ToolsView.dpToPx(8), ToolsView.dpToPx(2), ToolsView.dpToPx(8), 0);
        else if (viewWrapper instanceof DialogSheetWidget)
            vRecycler.getLayoutParams().height = ToolsView.dpToPx(320);
    }

    private void loadImages() {
        if (imagesLoaded) return;

        ToolsPermission.requestReadPermission(() -> {
            imagesLoaded = true;
            String[] projection = new String[]{MediaStore.Images.ImageColumns.DATA};
            Cursor cursor = SupAndroid.appContext.getContentResolver().query(
                    MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                    projection,
                    null,
                    null,
                    MediaStore.Images.ImageColumns.DATE_TAKEN + " DESC");

            while (cursor.moveToNext()) adapter.add(new CardImage(new File(cursor.getString(0))));
        }, () -> {
            ToolsToast.show(SupAndroid.TEXT_ERROR_PERMISSION_READ_FILES);
            hide();
        });


    }

    private void loadLink(String link) {
        WidgetProgressTransparent progress = ToolsView.showProgressDialog();
        Debug.log(" > " + link);
        ToolsNetwork.getBytesFromURL(link, bytes -> {
            progress.hide();

            if (!ToolsBytes.isImage(bytes)) {
                ToolsToast.show(SupAndroid.TEXT_ERROR_CANT_LOAD_IMAGE);
                return;
            }

            onSelected(bytes);
        });
    }

    private void showLink() {
        new WidgetField()
                .setMediaCallback((w, s) -> {
                    w.hide();
                    loadLink(s);
                })
                .enableFastCopy()
                .setHint(SupAndroid.TEXT_APP_LINK)
                .setOnEnter(SupAndroid.TEXT_APP_CHOOSE,
                        (w, s) -> loadLink(s))
                .setOnCancel(SupAndroid.TEXT_APP_CANCEL)
                .asSheetShow();

    }

    private void openGallery() {
        ToolsBitmap.getFromGallery(file -> {
            try {
                onSelected(ToolsFiles.readFile(file));
            } catch (IOException e) {
                Debug.log(e);
                ToolsToast.show(SupAndroid.TEXT_ERROR_CANT_LOAD_IMAGE);
            }
        });
    }

    private void onSelected(byte[] bytes) {
        if (onSelected != null) onSelected.callback(WidgetChooseImage.this, bytes);
        hide();
    }

    //
    //  Setters
    //

    public WidgetChooseImage setOnSelected(Callback2<WidgetChooseImage, byte[]> onSelected) {
        this.onSelected = onSelected;
        return this;
    }

    public WidgetChooseImage setOnSelectedBitmap(Callback2<WidgetChooseImage, Bitmap> callback) {
        this.onSelected = (widgetChooseImage, bytes) -> callback.callback(WidgetChooseImage.this, ToolsBitmap.decode(bytes));
        return this;
    }

    //
    //  Card
    //

    private class CardImage extends Card {

        private final File file;
        private byte[] bytes;

        public CardImage(File file) {
            this.file = file;
        }

        @Override
        public int getLayout() {
            return R.layout.sheet_choose_image_card;
        }

        @Override
        public void bindView(View view) {
            ImageView vImage = view.findViewById(R.id.image);
            vImage.setOnClickListener(v -> {
                if (bytes != null) onSelected(bytes);
            });

            ImageLoader.load(new ImageLoaderFile(file)
                    .setImage(vImage)
                    .cashScaledBytes()
                    .onLoaded(bytes -> this.bytes = bytes)
                    .sizes(512, 512)
                    .options(ImageLoader.OPTIONS_RGB_565())
                    .cropSquare());
        }


    }

}

File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/widgets/WidgetChooseImage.kt
package com.sup.dev.android.views.widgets

import android.graphics.Bitmap
import android.provider.MediaStore
import android.support.v7.widget.GridLayoutManager
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.sup.dev.android.R
import com.sup.dev.android.app.SupAndroid
import com.sup.dev.android.tools.*
import com.sup.dev.android.views.support.adapters.recycler_view.RecyclerCardAdapter
import com.sup.dev.android.views.cards.Card
import com.sup.dev.android.views.dialogs.DialogSheetWidget
import com.sup.dev.android.views.dialogs.DialogWidget
import com.sup.dev.java.libs.debug.err
import com.sup.dev.java.tools.ToolsBytes
import com.sup.dev.java.tools.ToolsFiles
import com.sup.dev.java.tools.ToolsNetwork
import com.sup.dev.java.tools.ToolsThreads
import java.io.File
import java.io.IOException


open class WidgetChooseImage : WidgetRecycler(R.layout.widget_choose_image) {

    private val DP = ToolsView.dpToPx(1).toInt()
    private val myAdapter: RecyclerCardAdapter = RecyclerCardAdapter()
    private val vEmptyText: TextView = findViewById(R.id.vEmptyText)
    private val vFabGalleryContainer: View = ToolsView.inflate(R.layout.z_fab)
    private val vFabLinkContainer: View = ToolsView.inflate(R.layout.z_fab)
    private val vFabGallery: ImageView = vFabGalleryContainer.findViewById(R.id.vFab)
    private val vFabLink: ImageView = vFabLinkContainer.findViewById(R.id.vFab)

    private var onSelected: (WidgetChooseImage, ByteArray) -> Unit = { widgetChooseImage, bytes -> }
    private var imagesLoaded: Boolean = false
    private var spanCount = 3

    init {
        vContainer.addView(vFabGalleryContainer)
        vContainer.addView(vFabLinkContainer)

        vEmptyText.text = SupAndroid.TEXT_ERROR_CANT_FIND_IMAGES

        (vFabLinkContainer.getLayoutParams() as ViewGroup.MarginLayoutParams).rightMargin = ToolsView.dpToPx(72).toInt()

        spanCount = if (ToolsAndroid.isScreenPortrait()) 3 else 6
        vRecycler.layoutManager = GridLayoutManager(view.context, spanCount)

        vFabGallery.setImageResource(R.drawable.ic_landscape_white_24dp)
        vFabLink.setImageResource(R.drawable.ic_insert_link_white_24dp)
        vFabGallery.setOnClickListener { v -> openGallery() }
        vFabLink.setOnClickListener { v -> showLink() }

        setAdapter<WidgetRecycler>(myAdapter)
    }

    override fun onShow() {
        super.onShow()
        loadImages()

        (vRecycler.layoutParams as ViewGroup.MarginLayoutParams).setMargins(0, 0, 0, 0)
        vRecycler.layoutParams.height = ViewGroup.LayoutParams.MATCH_PARENT

        if (viewWrapper is DialogWidget)
            (vRecycler.layoutParams as ViewGroup.MarginLayoutParams).setMargins(ToolsView.dpToPx(8).toInt(), ToolsView.dpToPx(2).toInt(), ToolsView.dpToPx(8).toInt(), 0)
        else if (viewWrapper is DialogSheetWidget)
            vRecycler.layoutParams.height = ToolsView.dpToPx(320).toInt()
    }

    private fun loadImages() {
        if (imagesLoaded) return

        ToolsPermission.requestReadPermission({
            imagesLoaded = true
            val projection = arrayOf(MediaStore.Images.ImageColumns.DATA)
            val cursor = SupAndroid.appContext!!.contentResolver.query(
                    MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                    projection, null, null,
                    MediaStore.Images.ImageColumns.DATE_MODIFIED + " DESC")

            while (cursor!!.moveToNext()) myAdapter.add(CardImage(File(cursor.getString(0))))

            myAdapter.clear()

            vEmptyText.visibility = if (myAdapter.isEmpty) View.VISIBLE else View.GONE
        }, {
            ToolsToast.show(SupAndroid.TEXT_ERROR_PERMISSION_FILES)
            hide()
        })


    }

    private fun loadLink(link: String) {
        val progress = ToolsView.showProgressDialog()
        ToolsNetwork.getBytesFromURL(link, onResult = {
            progress.hide()
            if (it == null || !ToolsBytes.isImage(it)) ToolsToast.show(SupAndroid.TEXT_ERROR_CANT_LOAD_IMAGE)
            else onSelected(it)
        })
    }

    private fun showLink() {
        WidgetField()
                .setMediaCallback { w, s ->
                    w.hide()
                    loadLink(s)
                }
                .enableFastCopy()
                .setHint(SupAndroid.TEXT_APP_LINK)
                .setOnEnter(SupAndroid.TEXT_APP_CHOOSE) { w, s -> loadLink(s) }
                .setOnCancel(SupAndroid.TEXT_APP_CANCEL)
                .asSheetShow()

    }

    private fun openGallery() {
        ToolsBitmap.getFromGallery({ bytes ->
            try {
                onSelected(bytes)
            } catch (e: IOException) {
                err(e)
                ToolsToast.show(SupAndroid.TEXT_ERROR_CANT_LOAD_IMAGE)
            }
        })
    }

    private fun onSelected(bytes: ByteArray) {
        onSelected.invoke(this, bytes)
        hide()
    }

    //
    //  Setters
    //

    fun setOnSelected(onSelected: (WidgetChooseImage, ByteArray) -> Unit): WidgetChooseImage {
        this.onSelected = onSelected
        return this
    }

    fun setOnSelectedBitmap(callback: (WidgetChooseImage, Bitmap) -> Unit): WidgetChooseImage {
        this.onSelected = { w, bytes -> callback.invoke(this, ToolsBitmap.decode(bytes)!!) }
        return this
    }

    //
    //  Card
    //

    private inner class CardImage(private val file: File) : Card() {

        override fun getLayout(): Int {
            return R.layout.sheet_choose_image_card
        }

        override fun bindView(view: View) {
            super.bindView(view)
            val vImage = view.findViewById<ImageView>(R.id.vImage)
            vImage.setOnClickListener { v -> onClick() }
            ToolsImagesLoader.load(file).size(512, 512).cropSquare().into(vImage)

            val index = adapter!!.indexOf(this)
            val arg = index % spanCount
            view.setPadding(if (arg == 0) 0 else DP, if (index < spanCount) 0 else DP, if (arg == spanCount - 1) 0 else DP, DP)

        }

        fun onClick() {
            val d = ToolsView.showProgressDialog()
            ToolsThreads.thread {
                try {
                    val bytes = ToolsFiles.readFile(file)
                    ToolsThreads.main {
                        d.hide()
                        onSelected(bytes)
                    }
                } catch (e: Exception) {
                    err(e)
                    ToolsThreads.main { d.hide() }
                    ToolsToast.show(SupAndroid.TEXT_ERROR_CANT_LOAD_IMAGE)
                }
            }
        }

    }

}
